Changelog linting

shuyama1 opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

This PR added a GHA for changelog unit tests and 3 more changelog linting rules:
1. Only one line is allowed per entry
2. changelog for new resource or datasource should be in format "`google_resource/datasource_name`"
3. changelog for enhancement or bug fix should be in format "package: change details"

No chnagelog error: https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/8485433892/job/23250199809

Some example errors: https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/8485475554/job/23250325859
<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:none
```

shuyama1 commented on tools/go-changelog/note.go:
```diff
@@ -48,7 +64,7 @@ func NotesFromEntry(entry Entry) []Note {
 			}
 
-			note = strings.TrimSpace(note)
+			// note = strings.TrimSpace(note)
```
We need to skip this here so that we can detect changelogs that contain multiple lines

melinath replied:
```suggestion
```

melinath commented on tools/go-changelog/cmd/changelog-pr-body-check/main.go:
```diff
@@ -57,31 +58,23 @@ func main() {
-			for _, t := range unknownTypes {
-				body += "\n* " + t
+	if errors := entry.Validate(); errors != nil {
+		log.Printf("error parsing changelog entry in %s: %s", entry.Issue, errors)
```
Since each of the errors will get printed out individually below, this output seems like it will just be a (less readable) duplicate... maybe it would be okay to leave out?

shuyama1 replied:
yeah, this makes sense

melinath commented on tools/go-changelog/cmd/changelog-pr-body-check/main.go:
```diff
@@ -57,31 +58,23 @@ func main() {
+			case changelog.EntryErrorNotFound:
+				body += fmt.Sprintf("\n\n Issue %d: It looks like no changelog entry is attached to this PR. Please include a release note block in the PR body, as described in https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/", i+1)
+			case changelog.EntryErrorUnknownTypes:
+				body += fmt.Sprintf("\n\n Issue %d: unknown changelog types %v \nPlease only use the types listed in https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/.", i+1, err.Details["type"].(string))
```
Rather than including each changelog inline here, what would you think about grouping the issues by changelog entry? So something like:

```
Changelog:
\`\`\`release-note:enhancement
compute engine: fixed an issue in google_resource resource 
\`\`\`

Issues:
  - invalid format. Please follow format in https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/#type-specific-guidelines-and-examples.
```

After typing this I realized that we probably currently only support one error per entry... so this would only make sense if we want to support multiple per entry.

This isn't a blocker; let me know what you think!

shuyama1 replied:
yes, we currently only support one error per entry. Plus, based on the current number of rules, I don't think we're gonna see multiple errors per changelog entry (to clarify, we do support multiple entries per PR though). But I could reform how the errors are reported. I can follow the example you have and make it more clear.

melinath reviewed (changes requested):
Could you modify the workflow for this to build the checker in a separate step so that the build output doesn't show up as part of the failure?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

shuyama1 commented:
I updated the PR description to include new GHA runs.

melinath commented on .github/workflows/changelog-checker.yml:
```diff
@@ -14,11 +14,22 @@ jobs:
           path: repo
+
+      - name: Set up Go
+        uses: actions/setup-go@v4
```
This should use a specific hash rather than just `v4`. I think it could be worth also rebasing or merging main, which already has actions/checkout pinned.

melinath commented on .github/workflows/unit-test-go-changelog.yml:
```diff
@@ -0,0 +1,29 @@
+  test:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v4
```
same comment here - these should be pinned to a specific hash.

melinath reviewed (approved):
LGTM apart from a couple minor changes to avoid scorecard violations

melinath reviewed (approved):
could you also make an invalid change just to double-check that the printing logic works as expected?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

shuyama1 commented:
> could you also make an invalid change just to double-check that the printing logic works as expected?

Tested the latest version against some invalid changelogs:

No chnagelog error: https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/8485433892/job/23250199809

Some example errors: https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/8485475554/job/23250325859
(with the following changelog entries)

```release-note:note
this changelog contains a newline

```

```release-note:feature
this is to add a feature (invalid type)
```

```release-note:new-resource
google_resource_without_backticks
```

```release-note:enhancement
added a new field to `google_resource` resource
```

```release-note:bug
compute engine: fixed an issue in `google_resource` resource
```

```release-note:new-datasource
`google_new_datasource_valid_format`
```

```release-note:enhancement
compute: added a new field to `google_resource` resource (valid changlog, no error's expected)
```

melinath reviewed (approved):
LGTM
